File: modules/zingo-wallet/android/src/main/java/expo/modules/zingowallet/ZingoWalletModule.kt
package expo.modules.zingowallet

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.lang.ref.WeakReference
import uniffi.zingo_wallet_ffi.Balance
import uniffi.zingo_wallet_ffi.HistoryEntry
import uniffi.zingo_wallet_ffi.HistoryKind
import uniffi.zingo_wallet_ffi.SyncListener
import uniffi.zingo_wallet_ffi.SyncPhase
import uniffi.zingo_wallet_ffi.SyncState
import uniffi.zingo_wallet_ffi.Wallet
import uniffi.zingo_wallet_ffi.WalletException

class ZingoWalletModule : Module() {
  private val wallet: Wallet
    get() {
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      return SharedWallet.get(File(context.noBackupFilesDir, "zingo-wallet"))
    }

  override fun definition() = ModuleDefinition {
    Name("ZingoWallet")
    Events("onSyncState")

    AsyncFunction("exists") Coroutine { -> coded { wallet.exists() } }
    AsyncFunction("create") Coroutine { -> coded { wallet.create() } }
    AsyncFunction("restore") Coroutine { seedPhrase: String, birthday: Long -> coded { wallet.restore(seedPhrase, birthday) } }
    AsyncFunction("load") Coroutine { -> coded { wallet.load() } }
    AsyncFunction("deleteWallet") Coroutine { -> coded { wallet.delete() } }
    AsyncFunction("balance") Coroutine { -> coded { wallet.balance().toMap() } }
    AsyncFunction("history") Coroutine { -> coded { wallet.history().map { it.toMap() } } }
    AsyncFunction("startSync") Coroutine { -> coded { wallet.startSync(SyncStateForwarder(this@ZingoWalletModule)) } }
    AsyncFunction("pauseSync") Coroutine { -> coded { wallet.pauseSync() } }
    AsyncFunction("resumeSync") Coroutine { -> coded { wallet.resumeSync() } }
  }
}

private object SharedWallet {
  private var wallet: Wallet? = null

  @Synchronized
  fun get(directory: File): Wallet = wallet ?: Wallet(directory.path, "https://zec.rocks:443").also { wallet = it }
}

private class SyncStateForwarder(module: ZingoWalletModule) : SyncListener {
  private val module = WeakReference(module)

  override fun onSyncState(state: SyncState) {
    runCatching { module.get()?.sendEvent("onSyncState", state.toMap()) }
  }
}

private inline fun <T> coded(block: () -> T): T =
  try {
    block()
  } catch (error: WalletException) {
    throw CodedException(error.code(), error.reason(), error)
  }

private fun Balance.toMap() = mapOf("spendableZats" to spendableZats, "unshieldedZats" to unshieldedZats)

private fun HistoryEntry.toMap() =
  mapOf("txid" to txid, "kind" to kind.jsName(), "zats" to zats, "timestampMs" to timestampMs, "pending" to pending)

private fun SyncState.toMap() =
  mapOf("phase" to phase.jsName(), "firstSyncComplete" to firstSyncComplete, "errorCode" to errorCode, "errorReason" to errorReason)

private fun HistoryKind.jsName() =
  when (this) {
    HistoryKind.RECEIVED -> "received"
    HistoryKind.SENT -> "sent"
    HistoryKind.SHIELDED -> "shielded"
  }

private fun SyncPhase.jsName() =
  when (this) {
    SyncPhase.IDLE -> "idle"
    SyncPhase.SYNCING -> "syncing"
    SyncPhase.SYNCED -> "synced"
    SyncPhase.PAUSED -> "paused"
    SyncPhase.FAILED -> "failed"
  }

File: modules/zingo-wallet/ios/ZingoWalletModule.swift
import ExpoModulesCore

private let walletDirectory = URL.applicationSupportDirectory.appending(path: "zingo-wallet", directoryHint: .isDirectory)

private let sharedWallet = Result<Wallet, Error> {
  let wallet = try Wallet(walletDir: walletDirectory.path, indexerUri: "https://zec.rocks:443")
  var directory = walletDirectory
  var backupExclusion = URLResourceValues()
  backupExclusion.isExcludedFromBackup = true
  try directory.setResourceValues(backupExclusion)
  return wallet
}

public final class ZingoWalletModule: Module {
  public func definition() -> ModuleDefinition {
    let syncStateForwarder = SyncStateForwarder(module: self)

    Name("ZingoWallet")
    Events("onSyncState")

    AsyncFunction("exists") { try await withWallet { try await $0.exists() } }
    AsyncFunction("create") { try await withWallet { try await $0.create() } }
    AsyncFunction("restore") { (seedPhrase: String, birthday: Int64) in
      try await withWallet { try await $0.restore(seedPhrase: seedPhrase, birthday: birthday) }
    }
    AsyncFunction("load") { try await withWallet { try await $0.load() } }
    AsyncFunction("deleteWallet") { try await withWallet { try await $0.delete() } }
    AsyncFunction("balance") { try await withWallet { try await $0.balance() }.dictionary }
    AsyncFunction("history") { try await withWallet { try await $0.history() }.map(\.dictionary) }
    AsyncFunction("startSync") { try await withWallet { try await $0.startSync(listener: syncStateForwarder) } }
    AsyncFunction("pauseSync") { try await withWallet { try await $0.pauseSync() } }
    AsyncFunction("resumeSync") { try await withWallet { try await $0.resumeSync() } }
  }
}

private final class SyncStateForwarder: SyncListener, @unchecked Sendable {
  private weak var module: ZingoWalletModule?

  init(module: ZingoWalletModule) {
    self.module = module
  }

  func onSyncState(state: SyncState) {
    module?.sendEvent("onSyncState", state.dictionary)
  }
}

private final class WalletErrorException: Exception, @unchecked Sendable {
  private let walletCode: String
  private let walletReason: String

  init(_ error: WalletError) {
    walletCode = error.code()
    walletReason = error.reason()
    super.init()
  }

  override var code: String { walletCode }
  override var reason: String { walletReason }
}

private func withWallet<T>(_ body: (Wallet) async throws -> T) async throws -> T {
  do {
    return try await body(sharedWallet.get())
  } catch let error as WalletError {
    throw WalletErrorException(error)
  }
}

private extension Balance {
  var dictionary: [String: Any] {
    ["spendableZats": spendableZats, "unshieldedZats": unshieldedZats]
  }
}

private extension HistoryEntry {
  var dictionary: [String: Any] {
    ["txid": txid, "kind": kind.jsName, "zats": zats, "timestampMs": timestampMs, "pending": pending]
  }
}

private extension SyncState {
  var dictionary: [String: Any?] {
    ["phase": phase.jsName, "firstSyncComplete": firstSyncComplete, "errorCode": errorCode, "errorReason": errorReason]
  }
}

private extension HistoryKind {
  var jsName: String {
    switch self {
    case .received: "received"
    case .sent: "sent"
    case .shielded: "shielded"
    }
  }
}

private extension SyncPhase {
  var jsName: String {
    switch self {
    case .idle: "idle"
    case .syncing: "syncing"
    case .synced: "synced"
    case .paused: "paused"
    case .failed: "failed"
    }
  }
}
